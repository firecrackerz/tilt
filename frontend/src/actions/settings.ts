import { Dispatch } from "redux";
import { IAction } from ".";
import { IEmailSettings, IEmailTemplates, ISettings } from "../../../types/settings";
import { api } from "../api";
import { notifyChangesSaved } from "./notify";
import { performRequest } from "./request";

/**
 * Settings redux action types.
 */
export enum SettingsAction {
  SetSettings = "set_settings",
}

/**
 * Creates a @see SettingsAction.SetSettings action.
 * @param settings The settings to set
 */
export const setSettings = (settings: ISettings): IAction<SettingsAction.SetSettings, ISettings> => ({
  type: SettingsAction.SetSettings,
  value: settings,
});

/**
 * Asynchronously fetches settings.
 */
export const fetchSettings = () => performRequest(async (dispatch: Dispatch) => {
  const settings = await api.getSettings();
  dispatch(setSettings(settings));
});

/**
 * Asynchronously update email settings.
 * @param settings The settings to update
 */
export const updateEmailSettings = (settings: Partial<IEmailSettings>) => performRequest(async (dispatch: Dispatch) => {
  await api.updateEmailSettings(settings);
  notifyChangesSaved()(dispatch);
});

/**
 * Asynchronously update email templates.
 * @param templates The templates to update
 */
export const updateEmailTemplates = (templates: Partial<IEmailTemplates>) => performRequest(async (dispatch: Dispatch) => {
  await api.updateEmailTemplates(templates);
  notifyChangesSaved()(dispatch);
});
