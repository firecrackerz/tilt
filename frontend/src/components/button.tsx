import styled from "styled-components";
import { borderRadius, transitionDuration } from "../config";
import { IThemeProps } from "../theme";

interface IButtonProps {
  primary?: boolean;
  fluid?: boolean;
  loading?: boolean;
}

/**
 * A clickable button.
 */
export const Button = styled.button<IButtonProps>`
  position: relative;
  top: 0px;

  display: inline-block;
  padding: 0.75rem 2rem;
  margin: 0rem;

  ${(props) => props.fluid && `
    width: 100%;
  `}

  border: none;
  border-radius: ${borderRadius};
  box-shadow: 0px 7px 10px rgba(0, 0, 0, 0.1);

  font-size: 0.8rem;
  font-weight: bold;
  text-transform: uppercase;
  cursor: pointer;

  background: #333;
  background-color: #333;
  color: white;

  transition-property: box-shadow, top, background;
  transition-duration: ${transitionDuration};

  &:hover {
    color: white;
    top: -3px;
    box-shadow: 0px 7px 15px rgba(0, 0, 0, 0.15);
  }

  ${(props: IThemeProps & IButtonProps) => props.primary && props.theme.colorGradientStart && `
    background: linear-gradient(to right, ${props.theme.colorGradientStart}, ${props.theme.colorGradientEnd});
  `}
`;
