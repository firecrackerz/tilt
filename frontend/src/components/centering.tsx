import styled from "styled-components";

/**
 * A centered container. Requires an @see OuterCenteredContainer around this one.
 */
export const InnerCenteredContainer = styled.div``;

/**
 * The outer container to center an @see InnerCenteredContainer
 */
export const OuterCenteredContainer = styled.div`
  position: relative;
  width: 100%;
  height: 100%;

  ${InnerCenteredContainer} {
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
  }
`;

/**
 * A page sized container.
 */
export const PageSizedContainer = styled.div`
  width: 100vw;
  height: 100vh;
  overflow: auto;
`;
