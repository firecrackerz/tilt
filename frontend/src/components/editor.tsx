import * as React from "react";
import MonacoEditor from "react-monaco-editor";

interface IEditorProps {
  language: string;
  value: string;
  onChange: (value: string) => any;
}

/**
 * A more or less rich text editor to edit text.
 */
export const Editor = ({ language, value, onChange }: IEditorProps) => (
  <MonacoEditor
    language={language}
    value={value}
    onChange={(text) => onChange(text)}
    width="100%"
    height="40vh"
    options={{
      autoClosingBrackets: "always",
      colorDecorators: true,
      minimap: {
        enabled: false,
      },
    }}
  />
);
