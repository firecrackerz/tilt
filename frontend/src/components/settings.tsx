import * as React from "react";
import { ConnectedEmailSettings } from "./email-settings";
import { Heading } from "./headings";

/**
 * A settings dashboard to configure all parts of tilt.
 */
export const Settings = () => (
  <>
    <Heading>Settings</Heading>
    <ConnectedEmailSettings />
  </>
);
