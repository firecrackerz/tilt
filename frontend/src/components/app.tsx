import * as React from "react";
import { useEffect } from "react";
import { connect } from "react-redux";
import { bindActionCreators, Dispatch } from "redux";
import { ThemeProvider } from "styled-components";
import { IFrontendSettings } from "../../../types/settings";
import { fetchSettings as fetchSettingsRaw } from "../actions/settings";
import { IState } from "../state";
import { ITheme } from "../theme";
import { ConnectedLoginSignupForm } from "./login-signup-form";

interface IAppProps {
  settings: IFrontendSettings;
  fetchSettings: typeof fetchSettingsRaw;
}

/**
 * The main app component.
 */
export const App = ({ settings, fetchSettings }: IAppProps) => {
  useEffect(() => {
    fetchSettings();
  }, []);

  const theme: ITheme = {
    colorGradientEnd: settings.colorGradientEnd,
    colorGradientStart: settings.colorGradientStart,
    colorLink: settings.colorLink,
    colorLinkHover: settings.colorLinkHover,
  };

  return (
    <ThemeProvider theme={theme}>
      <ConnectedLoginSignupForm />
    </ThemeProvider>
  );
};

const mapStateToProps = (state: IState) => ({
  settings: state.settings.data.frontend,
});

const mapDispatchToProps = (dispatch: Dispatch) => {
  return bindActionCreators({
    fetchSettings: fetchSettingsRaw,
  }, dispatch);
};

/**
 * The main app component, connected to the redux store.
 */
export const ConnectedApp = connect(mapStateToProps, mapDispatchToProps)(App);
