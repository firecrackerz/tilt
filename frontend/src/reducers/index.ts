import { combineReducers } from "redux";
import { IState } from "../state";
import { loginReducer } from "./login";
import { settingsReducer } from "./settings";
import { signupReducer } from "./signup";

/**
 * The app's root reducer.
 */
export const rootReducer = combineReducers<IState>({
  login: loginReducer,
  settings: settingsReducer,
  signup: signupReducer,
});
