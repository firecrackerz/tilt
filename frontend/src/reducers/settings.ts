import { IActionReturnTypes } from "../actions";
import * as settingsActions from "../actions/settings";
import { SettingsAction } from "../actions/settings";
import { IState } from "../state";

type IStateType = IState["settings"];
type IActionType = IActionReturnTypes<typeof settingsActions>;

const initialColor = "#333";

/**
 * The initial settings state.
 */
export const initialSettingsState: IStateType = {
  data: {
    frontend: {
      colorGradientEnd: initialColor,
      colorGradientStart: initialColor,
      colorLink: initialColor,
      colorLinkHover: initialColor,
      loginImage: "http://placehold.it/300x300",
      signupImage: "http://placehold.it/300x300",
    },
  },
  error: "",
  fetchInProgress: false,
};

/**
 * The settings reducer.
 * @param state The current state
 * @param action The current action
 */
export const settingsReducer = (state: IStateType = initialSettingsState, action: IActionType): IStateType => {
  switch (action.type) {
    case SettingsAction.StartFetchSettings:
      return {
        ...state,
        fetchInProgress: true,
      };

    case SettingsAction.ReceiveSettings:
      return {
        ...state,
        data: action.value,
      };

    case SettingsAction.FailReceiveSettings:
      return {
        ...state,
        error: action.value,
      };

    default:
      return state;
  }
};
