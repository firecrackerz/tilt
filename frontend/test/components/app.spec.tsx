import { shallow } from "enzyme";
import * as React from "react";
import { App } from "../../src/components/app";
import { initialSettingsState } from "../../src/reducers/settings";

describe("App", () => {
  it("renders", () => {
    const fetchRole = jest.fn();
    const fetchSettings = jest.fn();
    const app = shallow((
      <App settings={initialSettingsState.frontend} fetchSettings={fetchSettings} fetchRole={fetchRole} />
    ));

    expect(app).toMatchSnapshot();
  });
});
