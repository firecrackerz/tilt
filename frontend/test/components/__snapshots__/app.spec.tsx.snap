// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders 1`] = `
<ThemeProvider
  theme={
    Object {
      "colorGradientEnd": "#333",
      "colorGradientStart": "#333",
    }
  }
>
  <Component />
</ThemeProvider>
`;
